File: src/cat24c.cpp
#include "cat24c.hpp"

#include <string.h>

bool CAT24C::connected() {
    uint8_t b;
    return i2c_read_blocking(i2c, i2cAddress, &b, 1, false) != PICO_ERROR_GENERIC;
}

bool CAT24C::busy() {
    // When busy, the device essentially falls off the bus
    return !connected();
}

bool CAT24C::read(uint16_t address, uint8_t count, uint8_t *buffer) {
    uint8_t bytes[] = { address >> 8, address & 0xFF };
    if (i2c_write_blocking(i2c, i2cAddress, bytes, 2, false) == PICO_ERROR_GENERIC) return false;

    i2c_read_blocking(i2c, i2cAddress, buffer, count, false);

    return true;
}

bool CAT24C::write(uint16_t address, uint8_t count, const uint8_t *buffer) {
    // Adapted from Qwiic EEPROM Arduino library

    //Break the buffer into page sized chunks
    uint16_t recorded = 0;
    while (recorded < count)
    {
        //Limit the amount to write to either the page size or the Arduino limit of 30
        int amt_to_write = count - recorded;

        if (amt_to_write > 1)
        {
            //Check for crossing of a page line. Writes cannot cross a page line.
            uint16_t page_number_1 = (address + recorded) / PAGE_SIZE;
            uint16_t page_number_2 = (address + recorded + amt_to_write - 1) / PAGE_SIZE;
            if (page_number_2 > page_number_1)
                amt_to_write = (page_number_2 * PAGE_SIZE) - (address + recorded); //Limit the read amt to go right up to edge of page barrier
        }

        //See if EEPROM is available or still writing a previous request
        while (busy()) //Poll device
            sleep_us(100);          //This shortens the amount of time waiting between writes but hammers the I2C bus

        uint8_t write_buffer[count + 2];
        write_buffer[0] = (uint8_t)((address + recorded) >> 8);
        write_buffer[1] = (uint8_t)((address + recorded) & 0xFF);
        memcpy(write_buffer + 2, buffer, count);

        if (i2c_write_blocking(i2c, i2cAddress, write_buffer, count + 2, false) != count + 2) return false;

        recorded += amt_to_write;

        sleep_ms(PAGE_WRITE_MS); //Delay the amount of time to record a page
    }

    return true;
}

File: rust-hal/src/cat24c.rs
use alloc::vec;
use embedded_hal::blocking::{i2c::{Write, Read}, delay::DelayMs};

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Cat24CError {
    I2C,
}

pub struct Cat24C<I2CDevice: Write<Error = E> + Read<Error = E> + 'static, E, Delay: DelayMs<u8> + 'static> {
    address: u8,
    i2c: &'static mut I2CDevice,
    delay: &'static mut Delay,
}

impl<E, I2CDevice: Write<Error = E> + Read<Error = E> + 'static, Delay: DelayMs<u8> + 'static> Cat24C<I2CDevice, E, Delay> {
    pub fn new(
        address: u8,
        i2c: &'static mut I2CDevice,
        delay: &'static mut Delay,
    ) -> Self {
        Cat24C { address, i2c, delay }
    }

    #[allow(clippy::wrong_self_convention)] // &mut self required for SPI transmission
    pub fn is_connected(&mut self) -> bool {
        // TODO: Untested - I didn't have a Rev.1 to hand while writing this
        let mut buffer = [0; 1];
        self.i2c.read(self.address, &mut buffer[..]).is_ok()
    }

    #[allow(clippy::wrong_self_convention)] // &mut self required for SPI transmission
    pub fn is_busy(&mut self) -> bool {
        // When busy, the device essentially falls off the bus
        !self.is_connected()
    }

    pub fn read(&mut self, address: u16, bytes: &mut [u8]) -> Result<(), Cat24CError> {
        // Write the address we'd like to read from
        self.i2c.write(self.address, &[(address >> 8) as u8, (address & 0xFF) as u8])
            .map_err(|_| Cat24CError::I2C)?;

        // Read the desired number of bytes
        self.i2c.read(self.address, bytes).map_err(|_| Cat24CError::I2C)?;

        Ok(())
    }

    pub fn write(&mut self, address: u16, bytes: &[u8]) -> Result<(), Cat24CError> {
        const PAGE_SIZE: usize = 64;

        // Adapted from Qwiic EEPROM Arduino library
        // Break the buffer into page sized chunks
        let mut recorded = 0;
        while recorded < bytes.len()
        {
            // Limit the amount to write to the page size
            let mut amt_to_write = bytes.len() - recorded;

            if amt_to_write > 1
            {
                //Check for crossing of a page line. Writes cannot cross a page line.
                let page_number_1 = (address as usize + recorded) / PAGE_SIZE;
                let page_number_2 = (address as usize + recorded + amt_to_write - 1) / PAGE_SIZE;
                if page_number_2 > page_number_1 {
                    //Limit the read amt to go right up to edge of page barrier
                    amt_to_write = (page_number_2 * PAGE_SIZE) - (address as usize + recorded);
                }
            }

            // See if EEPROM is available or still writing a previous request
            while self.is_busy() {
                self.delay.delay_ms(1);
            }

            let mut write_buffer = vec![0; bytes.len() + 2];
            write_buffer[0] = ((address as usize + recorded) >> 8) as u8;
            write_buffer[1] = ((address as usize + recorded) & 0xFF) as u8;
            write_buffer[2..].copy_from_slice(bytes);

            self.i2c.write(self.address, &write_buffer[..])
                .map_err(|_| Cat24CError::I2C)?;

            recorded += amt_to_write;

            self.delay.delay_ms(5);
        }

        Ok(())
    }
}

